src/main/java/app/bpartners/api/service/event/InvoiceExportLinkRequestedService.java

```diff
@@ -0,0 +1,167 @@
+import app.bpartners.api.repository.UserRepository;
+import app.bpartners.api.repository.jpa.InvoiceJpaRepository;
+import app.bpartners.api.service.aws.S3Service;
+import app.bpartners.api.service.utils.TemplateResolverEngine;
+import jakarta.mail.internet.InternetAddress;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.LocalDate;
+import java.time.ZoneId;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Consumer;
+import lombok.AllArgsConstructor;
+import lombok.SneakyThrows;
+import org.jetbrains.annotations.NotNull;
+import org.springframework.stereotype.Service;
+import org.thymeleaf.context.Context;
+
+@Service
+@AllArgsConstructor
+public class InvoiceExportLinkRequestedService implements Consumer<InvoiceExportLinkRequested> {
+  public static final String INVOICE_EXPORT_LINK_REQUESTED_BODY = "invoice_export_link_requested";
+  private final FileZipper fileZipper;
+  private final Mailer mailer;
+  private final UserRepository userRepository;
+  private final InvoiceRepository invoiceRepository;
+  private final S3Service s3Service;
+  private final TemplateResolverEngine templateResolverEngine;
+  private final InvoiceJpaRepository invoiceJpaRepository;
+
+  @SneakyThrows
+  @Override
+  public void accept(InvoiceExportLinkRequested event) {
+    var accountId = event.getAccountId();
+    var providedFrom = event.getProvidedFrom();
+    var providedTo = event.getProvidedTo();
+
+    var fromLocalDatetime =
+        providedFrom == null ? now().withDayOfMonth(1).atStartOfDay() : providedFrom.atStartOfDay();
+    var toLocalDateTime =
+        providedTo == null
+            ? now().with(lastDayOfMonth()).atStartOfDay()
+            : providedTo.atStartOfDay();
+    var user = userRepository.getByIdAccount(accountId);
+    var userId = user.getId();
+    var from = fromLocalDatetime.atZone(ZoneId.systemDefault()).toInstant();
+    var to = toLocalDateTime.atZone(ZoneId.systemDefault()).toInstant();
+
+    var totalInvoices =
+        invoiceJpaRepository.countByIdUserAndCreatedDatetimeBetween(userId, from, to);
+    var nbPage = Math.max(1, (int) Math.ceil((double) totalInvoices / MAX_SIZE));
+    String zipFileId;
+    String htmlBody;
+
+    if (totalInvoices > 0) {
+      Path invoicesFiles = getTempDirectory();
+      for (int page = 0; page < nbPage; page++) {
+        List<Invoice> invoicePaginate =
+            invoiceRepository.findAllByIdUserAndCreateDateBetweenAndPaginate(
+                userId, from, to, page, MAX_SIZE);
+
+        var invoicePaginateFile = downloadInvoicesFiles(userId, invoicePaginate);
+
+        for (File invoice : invoicePaginateFile) {
+          Files.copy(invoice.toPath(), invoicesFiles, REPLACE_EXISTING);
+          Files.delete(invoice.toPath());
+        }
+      }
+      var invoicesZipFile = fileZipper.apply(List.of(invoicesFiles.toFile()));
+      zipFileId = randomUUID().toString();
```

Comment:
Why the declaration of zipFileId and its value assignation is separated ?

Reply:
there are 2 assignments of zipFileId on line 94 and 96